torch/csrc/jit/codegen/onednn/graph_helper.cpp

```diff
@@ -0,0 +1,519 @@
+  dnnl::graph::graph g{engineKind};
+
+  GRAPH_DEBUG("Constructing LLGA graph");
+  // TODO: select nodes in top-level block for now
+  for (auto* node : graph->block()->nodes()) {
+    auto op = createLlgaOp(node);
+    try {
+      g.add_op(op);
+      GRAPH_DEBUG("  Added node ", node->kind().toQualString());
+    } catch (std::exception& e) {
+      GRAPH_DEBUG(
+          "The backend failed to add node ", node->kind().toQualString());
+      g.add_op(makeWildcardOp(node).llgaOp());
+    }
+
+    for (Value* input : node->inputs()) {
+      tensorIdToValue_.emplace(input->unique(), input);
+    }
+  }
+
+  GRAPH_DEBUG("Get Partitions");
+  std::vector<dnnl::graph::partition> partitions = g.get_partitions(policy);
+  // excluded unsupported Wildcard partitions
+  for (auto& partition : partitions) {
+    if (partition.is_supported()) {
+      partitions_.push_back(partition);
+    }
+  }
+
+  GRAPH_DEBUG("  Got #partitions: ", partitions_.size());
+  for (size_t partId = 0; partId < partitions_.size(); partId++) {
+    for (auto opId : partitions_[partId].get_ops()) {
+      opToOwningPartition_.add(opId, partId);
+    }
+  }
+
+  // Scanning the graph again for post processing
+  for (auto* node : graph->block()->nodes()) {
+    mayAddListConstructIntoConcatPartition(node, opToOwningPartition_);
+  }
+}
+
+bool LlgaGraphHelper::isLlgaSubgraph(const Node* node) {
+  return node->hasAttribute(attr::Subgraph) &&
+      node->kind() == prim::oneDNNFusionGroup;
+}
+
+bool LlgaGraphHelper::shouldMerge(Node* toMerge, Node* subgraph) {
+  TORCH_CHECK(
+      isLlgaSubgraph(subgraph),
+      "The consumer node does not contain a subgraph");
+  if (!shouldConsiderForMerge(toMerge)) {
+    return false;
+  }
+  return opToOwningPartition_.get(toMerge) ==
+      opToOwningPartition_.get(subgraph);
+}
+
+// Except for conv & GEMMs, which should always be handled by oneDNN Graph,
+// only use single-op partitions for ops unsupported by NNC, or ops
+// that oneDNN executes faster. prim::ListConstruct is an exception, since
+// we simply want to fuse it with cat.
+bool isBetterSuitedForLLGA(NodeKind kindOfOp) {
+  if ((kindOfOp == aten::layer_norm) ||
+      (kindOfOp == aten::avg_pool2d) ||
+      (kindOfOp == aten::matmul) ||
+      (kindOfOp == aten::max_pool2d) ||
+      (kindOfOp == aten::conv2d) ||
+      (kindOfOp == aten::_convolution) ||
+      (kindOfOp == aten::mm) ||
+      (kindOfOp == aten::linear) ||
+      (kindOfOp == aten::cat) ||
+      (kindOfOp == prim::ListConstruct)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool LlgaGraphHelper::checkForSingleOpPartition(Node* node) {
+  if (opToOwningPartition_.has(node)) {
+    auto partitionId = opToOwningPartition_.get(node);
+    if (partitions_[partitionId].get_ops_num() == 1) {
+      auto kindOfNode = node->kind();
+      return isBetterSuitedForLLGA(kindOfNode);
+    } else {
+      // multi-op partition
+      return true;
```

Comment:
If a partition has multiple ops supported by oneDNN Graph, then it'd be more efficient for it to be handled by oneDNN Graph. Single-op oneDNN Graph partitions are only used for a few compute-intensive ops, because otherwise, there might be additional reorders, whose overhead might exceed the benefit of letting oneDNN Graph handle them.